mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp

```diff
@@ -74,24 +83,207 @@ static constexpr AmdArchInfo
+createTargetMachine(StringRef chip, StringRef featureString = "") {
+  static std::once_flag flag;
+  std::call_once(flag, [] {
+    LLVMInitializeAMDGPUTargetInfo();
+    LLVMInitializeAMDGPUTarget();
+    LLVMInitializeAMDGPUTargetMC();
+  });
+
+  Triple triple("amdgcn-amd-amdhsa");
+  std::string error;
+  const Target *target = TargetRegistry::lookupTarget(triple, error);
+  if (!target) {
+    llvm::errs() << "WARNING: Target registry lookup failed with error: "
+                 << error << ".\n";
+    return nullptr;
   }
-  StringRef chip = firstPart;
 
+  return std::unique_ptr<GCNTargetMachine>(static_cast<GCNTargetMachine *>(
+      target->createTargetMachine(triple, chip, featureString, {}, {}, {})));
+}
+
+AmdArchInfo fetchNativeArchInfo(const hipDeviceProp_t &prop) {
+  auto ret = lookupArchInfo(prop.gcnArchName); // get baseline
+
+  checkAndSetInfo("(HIP) minNumCU", ret.minNumCU, prop.multiProcessorCount);
+  checkAndSetInfo("(HIP) waveSize", ret.waveSize, prop.warpSize);
+  checkAndSetInfo("(HIP) totalSharedMemPerCU", ret.totalSharedMemPerCU,
+                  prop.maxSharedMemoryPerMultiProcessor);
+  checkAndSetInfo("(HIP) maxSharedMemPerWG", ret.maxSharedMemPerWG,
+                  prop.sharedMemPerBlock);
+
+  auto chip = std::get<0>(parseArchString(prop.gcnArchName));
+  if (auto tm = createTargetMachine(chip); !tm) {
+    llvm::errs() << "WARNING: Couldn't create target machine. Proceeding with "
+                    "HIP values.\n";
+  } else {
+    GCNSubtarget st(tm->getTargetTriple(), std::string(tm->getTargetCPU()),
+                    std::string(tm->getTargetFeatureString()), *tm);
+
+    checkAndSetInfo("(LLVM) numEUPerCU", ret.numEUPerCU, st.getEUsPerCU());
+    checkAndSetInfo("(LLVM) maxWavesPerEU", ret.maxWavesPerEU,
+                    st.getMaxWavesPerEU());
+    checkAndSetInfo("(LLVM) totalSGPRPerEU", ret.totalSGPRPerEU,
+                    st.getTotalNumSGPRs());
+    checkAndSetInfo("(LLVM) totalVGPRPerEU", ret.totalVGPRPerEU,
+                    st.getTotalNumVGPRs());
+    checkAndSetInfo("(LLVM) waveSize", ret.waveSize, st.getWavefrontSize());
+    checkAndSetInfo("(LLVM) totalSharedMemPerCU", ret.totalSharedMemPerCU,
+                    st.getLocalMemorySize());
+    checkAndSetInfo("(LLVM) maxSharedMemPerWG", ret.maxSharedMemPerWG,
+                    st.getAddressableLocalMemorySize());
+
+    auto features = ret.defaultFeatures;
+    if (st.hasAtomicFaddInsts()) {
+      features |= GemmFeatures::atomic_add;
+    } else {
+      features &= ~GemmFeatures::atomic_add;
+    }
+    if (st.hasAtomicBufferGlobalPkAddF16Insts() ||
+        st.hasAtomicBufferGlobalPkAddF16NoRtnInsts()) {
+      features |= GemmFeatures::atomic_add_f16;
+    } else {
+      features &= ~GemmFeatures::atomic_add_f16;
+    }
+    if (st.hasAtomicBufferPkAddBF16Inst()) {
+      features |= GemmFeatures::atomic_add_bf16;
+    } else {
+      features &= ~GemmFeatures::atomic_add_bf16;
```

Comment:
I am not sure if you need to explictly remove those. It would generate incorrect encoding. 
You need to remove only if they existed initially.